lib/matplotlib/backends/backend_pdf.py

```diff
@@ -1561,6 +1567,28 @@ def _writeImg(self, data, id, smask=None):
         if smask:
             obj['SMask'] = smask
         if mpl.rcParams['pdf.compression']:
+            if data.shape[-1] == 1:
+                data = data.squeeze(axis=-1)
+            img = Image.fromarray(data)
+            img_colors = img.getcolors(maxcolors=256)
+            if colors == 3 and img_colors is not None:
```

Comment:
Not originally your naming, but let's name it `color_channels` or `num_color_channels` to be precise.